### PR TITLE
Datasets and types: SQL datasets/type converters

### PR DESCRIPTION
**Pull request by aolfat**
# Description

Adds value converters for all the sql providers and rebuffs tests, Adds a common types test to implement for all the providers and reboots the unit tests to not use mock sql package and rely on integration tests for e2e flow

### Select type(s) of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (non-breaking change that modifies some code)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# Checklist:

- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have fixed any merge conflicts

## Summary by CodeRabbit

- **New Features**
  - Enhanced database integration with improved dataset writing and management across various providers, including PostgreSQL, MySQL, Snowflake, and BigQuery.
  - Introduced robust native data type conversion support for PostgreSQL, MySQL, Snowflake, and ClickHouse, ensuring accurate handling of diverse data formats.
  - Added new types and methods for managing writable datasets and handling batch operations in PostgreSQL, MySQL, Snowflake, and BigQuery.
  - Implemented a new function for sanitizing fully qualified object representations.

- **Tests**
  - Expanded test coverage to validate type conversions, batch operations, and edge case handling for various database providers, boosting overall system reliability.
  - Introduced comprehensive unit tests for the Converter functionality across different database providers, ensuring accurate type handling and conversion.

- **Refactor**
  - Updated internal type naming and registration mechanisms for increased consistency and improved data integration across multiple database providers.
  - Improved type definitions and method signatures for sanitizers, enhancing type clarity and safety.
  - Simplified iterator and data handling processes in SQL datasets.

- **Bug Fixes**
  - Enhanced error handling in conversion methods to ensure unsupported types are properly managed across different database providers.
  - Removed unnecessary pointer dereferencing functions to streamline code and improve maintainability.

### CODE REVIEW

File: delete_featureform_dynamo_tables.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
+#!/usr/bin/env python3
+import boto3
```
Comment (aolfat) [inline delete_featureform_dynamo_tables.py:2]: whoops don't commit this

File: fftypes/types.go [hunk 2]
```diff
@@ -335,16 +335,20 @@ type Schema struct {
 	columnSanitizer func(string) string
 }
 
+func (s *Schema) SetColumnSanitizer(sanitizer func(string) string) {
```
Comment (anthonylasso) [inline fftypes/types.go:338]: nit: why not just make the input a type?

File: provider/bigquery/dataset.go [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+// Err returns any error that occurred during iteration
+func (it *Iterator) Err() error {
+	return it.err
+}
+
+// Close closes the iterator
+func (it *Iterator) Close() error {
+	it.closed = true
```
Comment (anthonylasso) [inline provider/bigquery/dataset.go:141]: this just flips a flag, is there an actual close we need to call?

File: provider/bigquery/value_converter_test.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+		{"INT64", "INT64", types.Int64, false},
+		{"INTEGER", "INTEGER", types.Int64, false},
+		{"BIGINT", "BIGINT", types.Int64, false},
+		{"FLOAT64", "FLOAT64", types.Float64, false},
+		{"DECIMAL", "DECIMAL", types.Float64, false},
+		{"STRING", "STRING", types.String, false},
+		{"BOOL", "BOOL", types.Bool, false},
+		{"DATE", "DATE", types.Datetime, false},
+		{"DATETIME", "DATETIME", types.Datetime, false},
+		{"TIME", "TIME", types.Datetime, false},
```
Comment (anthonylasso) [inline provider/bigquery/value_converter_test.go:43]: should add some `expectErr=true` tests or to keep it clean remove the error logic here and add a separate expect errors test

File: provider/clickhouse/value_converter.go [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+// checkZeroTime returns a zero UTC time if the time is zero
+func checkZeroTime(t time.Time) time.Time {
+	if t.IsZero() {
+		return time.UnixMilli(0).UTC()
+	}
+	return t.UTC()
+}
+
+// deferencePointer handles potential pointer-to-pointer scenarios
+func deferencePointer(v interface{}) interface{} {
```
Comment (anthonylasso) [inline provider/clickhouse/value_converter.go:42]: nit: more idiomatic to use `any`

File: provider/clickhouse/value_converter.go [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+				NativeType: nativeType,
+				Type:       types.Int32,
+				Value:      int(intValue),
+			}, nil
+		}
+		return types.Value{
+			NativeType: nativeType,
+			Type:       types.UInt32,
+			Value:      value.(uint32),
+		}, nil
```
[the remark this answers is not in the document]
Comment (anthonylasso) [reply provider/clickhouse/value_converter.go:228]: @aolfat worth looking into.

File: provider/clickhouse_test.go [hunk 7]
```diff
@@ -53,6 +57,147 @@ func (ch *clickHouseOfflineStoreTester) CreateSchema(database, schema string) er
...
+	db *sql.DB
+}
+
+func (w WritableClickHouseDataset) WriteBatch(ctx context.Context, rows []types.Row) error {
+	if len(rows) == 0 {
+		return nil
+	}
+
+	schema := w.Schema()
+	if err != nil {
```
Comment (anthonylasso) [inline provider/clickhouse_test.go:71]: also not seeing where `err` gets declared? or is it meant to reference the package level err?

File: provider/clickhouse_test.go [hunk 8]
```diff
@@ -53,10 +57,152 @@ func (ch *clickHouseOfflineStoreTester) CreateSchema(database, schema string) er
...
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (anthonylasso) [reply provider/clickhouse_test.go:73]: noticed this as well, but it seems to point to a package level err?

File: provider/bigquery/dataset.go (hunk 3 above)
Comment (aolfat) [reply provider/bigquery/dataset.go:141]: ther's no close on the big query iter

File: fftypes/types.go [hunk 9]
```diff
@@ -328,23 +328,29 @@ type TypeConverter func(interface{}) (interface{}, error)
 
 type TypeConverterMapping map[string]TypeConverter
 
+type ColumnSanitizer func(string) string
```
Comment (simba-git) [inline fftypes/types.go:331]: This shouldn't be in Schema

File: provider/bigquery/value_converter.go [hunk 10]
```diff
@@ -59,7 +61,7 @@ func (c Converter) ConvertValue(nativeType types.NativeType, value any) (types.V
 		}, nil
 
 	// Floating point types
-	case "FLOAT64", "NUMERIC", "BIGNUMERIC":
+	case "FLOAT64", "DECIMAL":
```
[the remark this answers is not in the document]
Comment (simba-git) [reply provider/bigquery/value_converter.go:64]: +1

File: provider/bigquery/dataset.go [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+	return nil
+}
+
+// Next advances the iterator to the next row
+func (it *Iterator) Next() bool {
+	if it.closed {
```
Comment (simba-git) [inline provider/bigquery/dataset.go:147]: Should probably check if context is done as well

File: provider/bigquery/dataset.go (hunk 3 above)
Comment (simba-git) [reply provider/bigquery/dataset.go:141]: Actually this should also set an error so if someone calls Next the error will be IteratorClosed or something

File: provider/bigquery_test.go [hunk 12]
```diff
@@ -22,7 +22,10 @@ import (
 	"github.com/joho/godotenv"
 	"google.golang.org/api/option"
 
+	types "github.com/featureform/fftypes"
 	"github.com/featureform/logging"
+	bigquery2 "github.com/featureform/provider/bigquery"
```
[the remark this answers is not in the document]
Comment (aolfat) [reply provider/bigquery_test.go:27]: oh yeah, this was dumb, sorry

File: fftypes/types.go (hunk 9 above)
Comment (aolfat) [reply fftypes/types.go:331]: i'm thinking of not even making this its own type, wdyt?

File: provider/clickhouse/value_converter.go (hunk 5 above)
Comment (aolfat) [reply provider/clickhouse/value_converter.go:42]: this was from the old clickhouse code so I kept it, but i dont think it's necessary

File: provider/clickhouse/value_converter.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (aolfat) [reply provider/clickhouse/value_converter.go:228]: this is logic that whomever built clickhouse provider added

File: provider/clickhouse_test.go (hunk 7 above)
Comment (aolfat) [reply provider/clickhouse_test.go:71]: good catch! I refactored to not return err and forgot to delete

File: provider/clickhouse/value_converter.go [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+		}
+		return types.Value{
+			NativeType: nativeType,
+			Type:       types.Timestamp,
+			Value:      convertedValue,
+		}, nil
+	default:
+		// throw an error if the type is not supported
+		return types.Value{}, fferr.NewUnsupportedTypeError(typeStr)
+	}
```
[the remark this answers is not in the document]
Comment (anthonylasso) [reply provider/clickhouse/value_converter.go:265]: @aolfat might want to look into this

File: provider/snowflake/value_converter_test.go [hunk 14]
```diff
@@ -0,0 +1,283 @@
...
+
+	tests := []struct {
+		name       string
+		nativeType types.NativeType
+		value      interface{}
+		expected   types.Value
+		expectErr  bool
+	}{
+		// Integer tests
+		{"INTEGER nil", "INTEGER", nil, types.Value{NativeType: "INTEGER", Type: types.Int32, Value: nil}, false},
```
Comment (anthonylasso) [inline provider/snowflake/value_converter_test.go:99]: nit: wall of text, any chance you can break up this table test into type specific table tests.
